### PR TITLE
[MX-310] Improved Artist Rail interactions

### PR DESCRIPTION
**Pull request by ds300**
https://artsyproduct.atlassian.net/browse/MX-310

Well this ticket was a fair bit more than expected 🙈 Putting up as a draft because I want to try to get some tests implemented before the end of the day, and hopefully get it reviewed before I disappear for holiday.

The current UX is as follows:

- there are X artists on the rail
- they all have a 'follow' button regardless of whether you're following them or not
- when you tap the follow button the card disappears and is replaced with a new one, making sure it's not the artist that you just followed or an artist already in the rail, so there's always X unique artists on the rail

The new UX is as follows

- there are X artists on the rail
- they all have a button which is either in the 'following' or 'follow' state
- they all have a dismiss button in the top-right
- If you tap the follow/following button it ~has very complicated behaviour~ now just follows/unfollows the artist 💆 
- If you tap the dismiss button
  - The card goes into a 'spinning' loading state while we load another artist
  - We put the artist at the end of the list, then animate the dismissed artist away
  - (the reason we can't dismiss the artist card immediately then load the new artist afterwards is that it creates a janky experience for the user if they're at the end of the rail)
  - We need to make sure the new artist is not one that has already been dismissed during this session

Follow-up work in metaphysics:

1. Make `isFollowed` work on `'POPULAR'` homepage artist module.
2. Make `'SUGGESTED'` homepage artist module not include followed artists

# Gifs
_The 'based on' element is kinda messed up in these videos but fixed in the code base_

## Dismissing an artist card
dismiss-slide-middle

## Dismissing an artist card at the end of the rail
dismiss-slide-end

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Missing Test Files:

- `src/lib/Components/__tests__/Disappearable-tests.tsx`

If these files are supposed to not exist, please update your PR body to include "#skip_new_tests".</td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/lib/Components/Home/ArtistRails/ArtistRail.tsx [hunk 1]
```diff
@@ -167,64 +155,41 @@ const ArtistRail: React.FC<Props & RailScrollProps> = props => {
...
+      setArtists(_artists => _artists.concat([suggestion]))
+      await nextTick()
+    }
+    await artist._disappearable?.disappear()
+    setArtists(_artists => _artists.filter(a => a.internalID !== artist.internalID))
   }
 
-  // @ts-ignore STRICTNESS_MIGRATION
   const title = (): string => {
-    // TODO: Once Title is updated to styled-components, update the copy to spec
```
Comment (ashfurrow) [inline src/lib/Components/Home/ArtistRails/ArtistRail.tsx:216]: 🙌

File: src/lib/Components/Home/ArtistRails/ArtistRail.tsx [hunk 2]
```diff
@@ -233,42 +198,60 @@ const ArtistRail: React.FC<Props & RailScrollProps> = props => {
...
-        />
+        <SectionTitle title={title()} subtitle={subtitle()} />
       </Flex>
-      {renderModuleResults()}
+      <CardRailFlatList<SuggestedArtist>
+        listRef={listRef}
+        data={artists}
+        keyExtractor={artist => artist.id}
+        ItemSeparatorComponent={null}
+        // I noticed that sometimes FlatList seemed to get confused about where cards should be
```
Comment (ashfurrow) [inline src/lib/Components/Home/ArtistRails/ArtistRail.tsx:216]: 👍

File: src/lib/Components/Home/ArtistRails/ArtistRail.tsx [hunk 3]
```diff
@@ -233,42 +198,60 @@ const ArtistRail: React.FC<Props & RailScrollProps> = props => {
...
     fragment ArtistRail_rail on HomePageArtistModule {
       id
       key
       results {
-        id
-        internalID
-        ...ArtistCard_artist
+        # We get data for the artist card from three different queries, and we can't
+        # use data masking for two of them because relay GCs the masked data if not fetched via a fragment container
+        # so let's just not use masking for all three to keep things simple.
```
Comment (ashfurrow) [inline src/lib/Components/Home/ArtistRails/ArtistRail.tsx:253]: Mmmmm makes sense. I didn't realize Relay's GC was so keen!

File: src/lib/Scenes/Home/Components/ArtworkRail.tsx [hunk 4]
```diff
@@ -77,14 +71,17 @@ const ArtworkRail: React.FC<{ rail: ArtworkRail_rail } & RailScrollProps> = ({ r
           ListFooterComponent={() => <Spacer mr={2}></Spacer>}
           ItemSeparatorComponent={() => <Spacer mr={0.5}></Spacer>}
           showsHorizontalScrollIndicator={false}
-          data={rail.results}
+          data={rail.results as any /* TODO: fix this once MP's connections are non-nullable */}
```
Comment (ashfurrow) [inline src/lib/Scenes/Home/Components/ArtworkRail.tsx:74]: Aye, we've handled this here explicitly: 

https://github.com/artsy/eigen/pull/3271/files#diff-d8f4c64d1e1aa4e2813c0e2b1eed8fc8R62

File: src/lib/Components/Home/ArtistRails/ArtistRail.tsx [hunk 5]
```diff
@@ -43,98 +36,85 @@ interface Props extends ViewProperties {
...
+          basedOnArtistId: basedOnArtist.internalID,
+        }
+      )
+      const artist = result.artist?.related?.suggestedConnection?.edges?.[0]?.node ?? null
+      return (
+        artist && {
+          ...artist,
+          _disappearable: null,
+          // make the basedOn for this suggestion fall back to either the artist this was actually based on (if followed)
+          // or whatever _that_ artist suggestion was based on, if available. Transient basedOn!
```
Comment (ashfurrow) [inline src/lib/Components/Home/ArtistRails/ArtistRail.tsx:91]: Neat!

**Pull request merged**